webhooks/virtualmachine/validation/virtualmachine_validator.go

```diff
@@ -2112,23 +2113,127 @@ func (v validator) validateNextPowerStateChangeTimeFormat(
 	return allErrs
 }
 
+func (v validator) validateBootOptionsWhenPoweredOn(
+	ctx *pkgctx.WebhookRequestContext,
+	vm, oldVM *vmopv1.VirtualMachine) field.ErrorList {
+
+	var allErrs field.ErrorList
+	fieldPath := field.NewPath("spec", "bootOptions")
+
+	if !reflect.DeepEqual(oldVM.Spec.BootOptions, vm.Spec.BootOptions) {
+		allErrs = append(allErrs, field.Forbidden(fieldPath, updatesNotAllowedWhenPowerOn))
+	}
+
+	return allErrs
+}
+
 func (v validator) validateBootOptions(
 	ctx *pkgctx.WebhookRequestContext,
-	vm *vmopv1.VirtualMachine) field.ErrorList {
+	vm, oldVM *vmopv1.VirtualMachine) field.ErrorList {
 
 	var allErrs field.ErrorList
+	fieldPath := field.NewPath("spec", "bootOptions")
 
-	if vm.Spec.BootOptions != nil {
-		fieldPath := field.NewPath("spec", "bootOptions")
+	bootOptions := vm.Spec.BootOptions
+	if bootOptions != nil {
+		if bootOptions.BootOrder != nil {
+			create := oldVM == nil
+			allErrs = append(allErrs, v.validateBootOrder(ctx, vm, create)...)
+		}
 
+		// Both fields are required.
 		if vm.Spec.BootOptions.BootRetryDelay != nil && vm.Spec.BootOptions.BootRetry != vmopv1.VirtualMachineBootOptionsBootRetryEnabled {
-			allErrs = append(allErrs, field.Required(fieldPath.Child("bootRetry"), "when setting bootRetryDelay"))
+			allErrs = append(
+				allErrs,
+				field.Required(
+					fieldPath.Child("bootRetry"),
+					"bootRetry must be set when setting bootRetryDelay",
+				))
 		}
 
+		// EFISecureBoot can only be used in conjunction with EFI firmware
 		if vm.Spec.BootOptions.EFISecureBoot == vmopv1.VirtualMachineBootOptionsEFISecureBootEnabled &&
 			vm.Spec.BootOptions.Firmware != vmopv1.VirtualMachineBootOptionsFirmwareTypeEFI {
 
-			allErrs = append(allErrs, field.Forbidden(fieldPath.Child("efiSecureBoot"), "when image firmware is not EFI"))
+			allErrs = append(
+				allErrs,
+				field.Forbidden(
+					fieldPath.Child("efiSecureBoot"),
+					"cannot set efiSecureBoot when image firmware is not 'efi'",
+				))
+		}
+	}
+
+	return allErrs
+}
+
+func (v validator) validateBootOrder(
+	_ *pkgctx.WebhookRequestContext,
+	vm *vmopv1.VirtualMachine,
+	create bool) field.ErrorList {
+
+	var allErrs field.ErrorList
+	fieldPath := field.NewPath("spec", "bootOptions", "bootOrder")
+
+	if create {
+		// We do not allow for specifying bootOrder on create.
+		allErrs = append(allErrs, field.Forbidden(fieldPath, "when creating a VM"))
+	} else {
+		// Validate that the devices listed in bootOrder are valid. That is,
+		// validate that the device type is one of the valid options and that
+		// the device exists.
+		for _, bd := range vm.Spec.BootOptions.BootOrder {
+			switch bd.Type {
+			case vmopv1.VirtualMachineBootOptionsBootableNetworkDevice:
+				// Validate that the named network interface is present.
+				found := slices.ContainsFunc(vm.Spec.Network.Interfaces, func(iface vmopv1.VirtualMachineNetworkInterfaceSpec) bool {
```

Comment:
Just double checking - could `vm.Spec.Network` be nil here?

Reply:
i don't think so. we mutate that field on create when it isn't provided and if it doesn't have the v1alpha1 NoNic annotation present. but we don't allow for setting boot order on create. on update, we don't allow for removing the nic. however, since it is a pointer field there should be a check. adding.